drivers/sensor/ltrf216a/ltrf216a.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2023 Tridonic
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ltr_f216a
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util_macro.h>
+
+LOG_MODULE_REGISTER(ltrf216a, CONFIG_SENSOR_LOG_LEVEL);
+
+/*
+ * Driver for I2C illuminance sensor LiteOn LTR-F216A.
+ *
+ * Datasheet:
+ * <https://optoelectronics.liteon.com/upload/download/DS86-2019-0016/LTR-F216A_Final_DS_V1.4.PDF>
+
+ * 7bit Address 0x53
+ * 8bit Address 0xA6 read
+ * 8bit Address 0xA7 write
+
+ * NOT IMPLEMENTED:
+ * - Interrupt
+ * - Modifying Gain (using default x3)
+ * - Modifying Resolution (using default 100ms)
+ * - Modifying Measurement Rate (using default 100ms)
+ * - Modifying Window Factor (using default 1)
+ */
+
+#define LTRF216A_ALS_RESET_MASK  BIT(4)
+#define LTRF216A_ALS_ENABLE_MASK BIT(1)
+
+#define LTRF216A_ALS_DATA_STATUS BIT(3)
+
+/* Part Number ID 7:4 0b1011 = 0xB */
+/* Revision ID 3:0 0b0001 = 0x1*/
+#define LTRF216A_PART_ID_VALUE 0xB1
+
+#define LTRF216A_MAIN_CTRL        0x00
+#define LTRF216A_ALS_MEAS_RES     0x04
+#define LTRF216A_ALS_GAIN         0x05
+#define LTRF216A_PART_ID          0x06
+#define LTRF216A_MAIN_STATUS      0x07
+#define LTRF216A_ALS_CLEAR_DATA_0 0x0a
+#define LTRF216A_ALS_CLEAR_DATA_1 0x0b
+#define LTRF216A_ALS_CLEAR_DATA_2 0x0c
+#define LTRF216A_ALS_DATA_0       0x0d
+#define LTRF216A_ALS_DATA_1       0x0e
+#define LTRF216A_ALS_DATA_2       0x0f
+#define LTRF216A_INT_CFG          0x19
+#define LTRF216A_INT_PST          0x1a
+#define LTRF216A_ALS_THRES_UP_0   0x21
+#define LTRF216A_ALS_THRES_UP_1   0x22
+#define LTRF216A_ALS_THRES_UP_2   0x23
+#define LTRF216A_ALS_THRES_LOW_0  0x24
+#define LTRF216A_ALS_THRES_LOW_1  0x25
+#define LTRF216A_ALS_THRES_LOW_2  0x26
+
+#define LTRF216A_ALS_READ_DATA_DELAY_MS 20
+
+#define LTRF216A_WIN_FAC 1
+
+#define LTRF216A_NUMBER_DATA_REGISTERS 3
+
+struct ltrf216a_data {
+	uint8_t sample[3];
+};
+
+struct ltrf216a_config {
+	struct i2c_dt_spec i2c;
+};
+
+static int ltrf216a_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct ltrf216a_data *drv_data = dev->data;
+	const struct ltrf216a_config *config = dev->config;
+	uint8_t status;
+	int result;
+	uint8_t value;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_LIGHT);
+
+	drv_data->sample[0] = 0U;
+	drv_data->sample[1] = 0U;
+	drv_data->sample[2] = 0U;
+
+	value = LTRF216A_ALS_ENABLE_MASK;
+	result = i2c_burst_write_dt(&config->i2c, LTRF216A_MAIN_CTRL, &value, 1);
+	if (result != 0) {
+		LOG_ERR("ltfr216a: enable failed");
+		return result;
+	}
+
+	while ((result = i2c_burst_read_dt(&config->i2c, LTRF216A_MAIN_STATUS, &status, 1)) == 0) {
```

Comment:
Measurement (with default values) should be done after 100ms.

linux kernel uses: 
<https://github.com/torvalds/linux/blob/6465e260f48790807eef06b583b38ca9789b6072/drivers/iio/light/ltrf216a.c#L216-L219>

I can fetch samples without a delay, but only if fetch isn't called faster than 100ms (i can be faster than that with shell's "sensor get").

Should i assume that? Should i guard against that? Or should i keep the delay?

Reply:
Thanks for looking into this.
I think a fixed number of retries would be best (based on the worst case delay).  
As you say the delay is still needed in case the fetch is called too soon, and forcing a delay when it might be needed is a bit unpleasant.

Reply:
@aasinclair I tried and I am failing to get a retry mechanism working with this chip. I would for the moment rather not retry and just return an error if called too quickly. OK?

Reply:
i have pushed the solution without retry now.